benchmarks/lockhammer/scripts/sweep.sh

```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+# Copyright (c) 2017, The Linux Foundation. All rights reserved.
+# 
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#     * Redistributions of source code must retain the above copyright
+#       notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+#       copyright notice, this list of conditions and the following
+#       disclaimer in the documentation and/or other materials provided
+#       with the distribution.
+#     * Neither the name of The Linux Foundation nor the names of its
+#       contributors may be used to endorse or promote products derived
+#       from this software without specific prior written permission.
+# 
+# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+# BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+# BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+# IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+cores=$(grep -c "^processor" /proc/cpuinfo)
+cores_q1=$(($cores / 4))
+cores_q2=$(($cores / 2))
+cores_q3=$(($cores_q1 + $cores_q2))
+cores_all="`seq 24` `seq 8 8 $(($cores))` $cores_q1 $cores_q2 $cores_q3 $cores"
```

Comment:
Should we modify this line to:
cores_all="`seq 24` `seq 8 8 64` `seq 64 16 $(($cores))` $cores_q1 $cores_q2 $cores_q3 $cores"

By default, 256 core system would have these test points, otherwise entire sweep time would be too long for higher core count system, single stepping from 1 to 24 should be enough to cover low and medium core count system:
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
32
40
48
56
64
80
96
112
128
144
160
176
192
208
224
240
256

Reply:
This is starting to get a bit complicated.  Would it make more sense to have the acquires-per-thread count adjust based on number of threads?  Since acquires-per-thread is held constant the total number of acquires increases linearly with number of threads.  Since many of these operations saturate at either the critical section time or synchronization overhead time we don't see any scaling at very high core counts.  So the total work increase with the addition of cores but the aggregate SoC performance stays the same causing runtimes to increase linearly with core counts at the higher core.

By changing the acquires per thread at higher core counts (say, > 64) to either hold constant the total acquires or increase at a flatter rate we could have more impact on the total runtime than removing some of the datapoints.  It may also improve responsiveness for the very high-core count systems by reducing the maximum time that any one instance can run which might prevent the hangups you've seen.

Reply:
I think that is a good starting point.  Lets divide the number of iterations down to a minimum of 1000, and at that point I think this commit and issue are in a good enough state to get it merged in.  We can work on polishing the code through other issues at a later time.  I think we can merge this in by tomorrow with this last change.

Reply:
Yes, with current implementation, the testing time increases super-linearly with core count. We need at least 15 ~ 30 minutes to cover even one core point above 96 cores. That's the reason we want to skip more core points beyond 64.

I agree we should keep the total lock iteration a fixed value, and each thread get its own adjustable value by dividing this total number with thread count. So that we can make each core point's test time relatively the same (or slightly increase), and reduce the time needed for higher core count.

For example, if 1000 iterations is the minimum number for a single thread, 256 core system would execute 256 * 1000 = 256k iterations in total. If we use this as the fixed total iteration for 64 ~ 256 threads, 64 threads would use 4k per thread iteration. 128 threads would use 2k per thread. For < 64 threads, e.g. 32 threads would use 8k, 16t => 16k, 8t => 32k, 4t or less == 50k

And per thread iteration will max out at 50k. How about it?

Reply:
However, even with fixed total iteration solution, we don't need to test every core count. Beyond 64 cores, 16 core step should be enough, 8 core step is overkill.

Reply:
My only concern is that for certain topologies some useful information might be thrown out.  For example, the thread pinning is setup such that on Skylake thread siblings are populated before full cores (so a 2 core run would be both thread siblings on the first physical core).  This means that for multi-socket Skylake the addition of the first core on the second socket comes at the halfway point in the logical core count.  If we're only increasing by 16 threads we might skip over some interesting information about the transition from the last core on the first socket to the first core on the second socket.

Reply:
We covered that by "seq 24" which will test each core count until 24, which will reveal topologies like hyperthread or arm cluster. This value is easily tunable, we can change this single value to 48 and it will cover all cores until 48. If we change it to max cores, it will cover every core point. However, it is not easy to remove certain core point if we test every core point by default.

Another reason is test time, if by default we sweep all core counts until the max config, 128 core system needs several hours even for a single lock test.

Reply:
The change to skipping 8 and the clamped acquires per thread means that with the changes already in the PR have a reduction in runtime by a factor equal to the number of cores (N/8 from acquire count reduction and then 8 from the skipping 8 cores) at high core counts.  So 64x faster at 64 cores and 128x faster at 128 cores.  Further raising the core skipping to 16 cores above 64 cores only gives an additional improvement of 1.12x for a 128 core system (24 + 13 vs 24 + 5 + 4) while increasing complexity of the run script and halving the number of data points in the high-contention region of operation.

Reply:
I think this looks good as it is, no need to bike shed to death an example script that I think many will modify on their own anyways.